src/core/apparmor-setup.c

```diff
@@ -0,0 +1,98 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <errno.h>
+#include <unistd.h>
+
+#include "fd-util.h"
+#include "fileio.h"
+#include "log.h"
+#include "macro.h"
+#include "apparmor-setup.h"
+#include "apparmor-util.h"
+#include "string-util.h"
+
+#if HAVE_APPARMOR
+#include <sys/apparmor.h>
+#endif
+
+#define APPARMOR_PROFILE_NAME "systemd"
+#define APPARMOR_PROFILE_ETC_PATH "/etc/systemd/profile.apparmor"
+#define APPARMOR_PROFILE_USR_PATH "/usr/lib/systemd/profile.apparmor"
+
+int mac_apparmor_setup(void) {
+#if HAVE_APPARMOR
+        int r = -1;
+        _cleanup_close_ int fd = -1;
+        aa_kernel_interface *interface = NULL;
+        const char* profile_file = APPARMOR_PROFILE_ETC_PATH;
+        _cleanup_free_ char *current_profile = NULL;
+
+        if (!mac_apparmor_use()) {
+                log_debug("AppArmor either not supported by the kernel or disabled.");
+                return 0;
+        }
+
+        r = read_one_line_file("/proc/self/attr/apparmor/current", &current_profile);
+        if (r < 0) {
+                if (errno == ENOENT) {
+                        r = read_one_line_file("/proc/self/attr/current", &current_profile);
+                        if (r < 0) {
+                                log_warning_errno(errno, "Failed to get the current AppArmor profile of systemd from /proc/self/attr/current: %m");
+                                return 0;
+                        }
+                } else {
+                        log_warning_errno(errno, "Failed to get the current AppArmor profile of systemd from /proc/self/attr/apparmor/current: %m");
+                        return 0;
+                }
+        }
+        if (!streq(current_profile, "unconfined")) {
+                log_debug("Systemd is already confined in an AppArmor profile.");
+                return 0;
+        }
+
+        if (access(profile_file, F_OK) < 0) {
+                profile_file = APPARMOR_PROFILE_USR_PATH;
+                if (access(profile_file, F_OK) < 0) {
+                        log_debug("AppArmor profile file neither found at path " APPARMOR_PROFILE_ETC_PATH " nor " APPARMOR_PROFILE_USR_PATH ".");
+                        return 0;
+                }
+        }
+
+        fd = open(profile_file, O_RDONLY|O_CLOEXEC);
```

Comment:
nah, please don't test first and open second. Just open right-away and if you get ENOENT try the next file until you find it.

Also consider using the FOREACH_STRING macro, i.e. something like this:

```c
FOREACH_STRING(fn, "/etc/systemd/profile.apparmor", "/usr/lib/systemd/profile.apparmor") {
        fd = open(fn, O_RDONLY|O_CLOEXEC);
        if (fd >= 0) 
                break;
        if (errno != ENOENT) 
                log_warning_errno(errno, "Failed to open %s, ignoring: %m");
}
if (fd < 0) {
        log_warning("Unable to find suitable AppArmor profile, yadda yadda, not loading profile.");
        return 0;
}
```

You get the idea